kep/205-preemption-toleration/README.md

```diff
@@ -0,0 +1,215 @@
+
+To achieve this, `PriorityClass` with a high priority value would be enough. But just giving high priority value causes preemption on any pods with lower priority values.  Sometimes, users don't want to do it.  In such a case, the job's priority is essentially low (i.e. the job is expected to run only when the cluster has a vacancy).  However, once scheduled, users don't want the pod to be preempted because the pod is difficult to export/import its state.
+
+On the other hand, from the cluster administrative view, unconditional non-preempted pods are not desirable because there are system-critical, node-critical pods in the cluster. So, the cluster administrator would like to set the minimum priority value for such a priority class.  For example, assume there are three priority classes: system-critical, high and low.  Then, the cluster administrator would like to introduce `low-non-preempted` priority class that can not be preempted by high but can be preempted by system-critical.
+
+```yaml
+  system-critical: 10000
+             high:  9000
+              low:  8000
+low-non-preempted:  8000 # with minimumPreemptablePriority=10000
+                        # (i.e. high can't preempt this, but system-critical can preempt this)
+```
+
+### Guarantee to running at least N minutes even in lower priority
+
+This is a typical use-case in machine learning job. As described above, job would need check-pointing. However, most machine learning jobs are iteration-based process, called _epochs_. So, if the cluster is high load and preemption always happens before the first epoch is finished,  the machine learning jobs can never make any progress and computational resources consumed by the job came to nothing in this case.
+
+So, cluster administrators would like to exempt pods that do not finish the first epoch from preemption even if the job has lower priority. But, it is hard to monitor when the first epoch is finished outside of the job. Thus, a minimum running time is useful practically.
+
+For example, cluster administrator would introduce several priority classes with the same priority value for introducing several minimum running time guarantee categories like below:
+
+```yaml
+    system-critical: 10000
+                low:  8000  # no minimum running time guarantee
+          low-10min:  8000  # 10 min minimum running time guarantee against 8000 < p < 10000
+          low-30min:  8000  # 30 min minimum running time guarantee against 8000 < p < 10000
+```
+
+## Design Details
+
+### Preemption Toleration API
+
+Preemption toleration policy is defined in this struct type.
+
+```go
+type PreemptionToleration struct {
+  // MinimumPreemptablePriority specifies the minimum priority value that can preempt this priority class.
+  // If not set, pods that have a higher priority value can preempt it (default).
```

Comment:
To standardize the defaulting logic, _ideally_ we should default the value to the PC's priority value. (but not 100% sure it's feasible, let's state it this way for now)

```
// It defaults to the PriorityClass's priority value if not set, which means pods that have a higher priority value can preempt it.
```

Reply:
OK. It's a nit but, the default value would be `PriorityClass's priority value + 1` because this is `minimumPreemptableValue` . I fill fix it with this.

Reply:
done in https://github.com/kubernetes-sigs/scheduler-plugins/pull/205/commits/d2ac06178e08980c69073dc4ab34cb1a1dad639e

Reply:
SG.